src/value.rs

```diff
@@ -186,3 +189,267 @@ impl From<Bytes> for UIntValue {
         Self::try_from(value.0.as_ref()).expect("At most 32 bytes")
     }
 }
+
+/// Various value constructors.
+pub trait ValueConstructible: Sized + From<Option<Self>> + From<bool> + From<UIntValue> {
+    /// Create the unit value.
+    fn unit() -> Self;
+
+    /// Create the left value `Either::Left(inner)`.
+    fn left(inner: Self) -> Self;
+
+    /// Create the right value `Either::Right(inner)`.
+    fn right(inner: Self) -> Self;
+
+    /// Create the product value `(left, right)`.
+    fn product(left: Self, right: Self) -> Self;
+
+    /// Create an array from the given `elements`.
+    ///
+    /// ## Nonemptiness
+    ///
+    /// There must be at least one element.
+    ///
+    /// ## Errors
+    ///
+    /// Returns [`None`] if there are no elements.
+    // TODO: Enforce the nonemptiness of the iterator through the type system?
+    fn array<I: IntoIterator<Item = Self>>(elements: I) -> Option<Self>;
+
+    /// Create `bound`ed list from the given `elements`.
+    ///
+    /// ## Boundedness
+    ///
+    /// There must be fewer `elements` than the given `bound`.
+    /// The `bound` is an exclusive upper bound on the number of `elements`.
+    ///
+    /// ## Errors
+    ///
+    /// Returns [`None`] if there are too many elements.
+    // TODO: Enforce the boundedness through the type system?
+    fn list<I: IntoIterator<Item = Self>>(elements: I, bound: NonZeroPow2Usize) -> Option<Self>;
+}
+
+/// Simfony value.
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub enum Value {
+    /// Unit value.
+    Unit,
+    /// Left value.
+    Left(Arc<Self>),
+    /// Right value.
+    Right(Arc<Self>),
+    /// Product value.
+    Product(Arc<Self>, Arc<Self>),
+    /// Option value.
+    Option(Option<Arc<Self>>),
+    /// Boolean value.
+    Boolean(bool),
+    /// Unsigned integer.
+    UInt(UIntValue),
+    /// Nonempty array of values.
+    Array(Arc<[Self]>),
+    /// Bounded list of values.
+    List(Arc<[Self]>, NonZeroPow2Usize),
```

Comment:
In 892062c29dcc3787f20539b31a7bc153c29ae507:

This `eum Value` allows the user to construct invalid values for the `List` and `Array` variants. That may be okay (or at least, too annoying to prevent) but we should be aware of it. The "correct" way to handle this is to move the enum into an `enum ValueInner` and have the public interface go through an opaque newtype around that with constructors that enforce validity.

Reply:
Preventing invalid array or list values would be annoying because we still want to do pattern matching. For now, I will add a comment that this is an issue.

We could add methods like `as_array()` and `as_list()` that inspect the value instead of doing pattern matching. There are not so many value variants.